postproc/postproc/traceDB.py

```diff
@@ -0,0 +1,360 @@
+        """
+        ss = f"name: {self.name}\n"
+        ss += f"unit: {self.unit}\n"
+        ss += f"multi: {self.multi}\n"
+        if raw_trace_idx:
+            ss += f"raw_traces:\n{next(iter(self.raw_traces, raw_trace_idx))}\n"
+        else:
+            ss += f"raw_traces:\n{self.raw_traces}\n"
+        ss += f"refined_traces:\n{self.refined_traces}"
+
+        return ss
+
+    def append_to_raw_trace(self, file_path, val):
+        """Service function to append to a raw_trace"""
+        val = val * self.multi
+        if file_path in self.raw_traces:
+            self.raw_traces[file_path].append(val)
+        else:
+            self.raw_traces[file_path] = [val]
+
+    def add_raw_trace(self, file_path, trace):
+        self.raw_traces[file_path] = self.multi * trace
+
+    def derive_raw_data(self, traces):
+        """Derive the current self.raw_traces from other traces
+
+        Using the other traces provided we evaluate the expression provided in derivation_params.
+        derivation_params has the following structure:
+
+        [operation, [trace_names], other_arguments]
+
+        and the derivation is computed so that:
+
+        operation(*[i for i in traces[trace_names]], *other_arguments)
+
+        Args:
+            - traces (dict): Complete dictionary of traces from TraceDB
+        """
+        if not self.derivation_params:
+            return
+
+        [op, root_traces, *args] = self.derivation_params
+
+        for file in traces[root_traces[0]].raw_traces:
+            input_traces = [traces[i].raw_traces[file] for i in root_traces]
+
+            if hasattr(Utils, op) and callable(func := getattr(Utils, op)):
+                derived_trace = func(*input_traces, *args)
+            elif hasattr(numpy, op) and callable(func := getattr(numpy, op)):
+                derived_trace = func(*input_traces, *args)
+            elif hasattr(scipy, op) and callable(func := getattr(scipy, op)):
+                derived_trace = func(*input_traces, *args)
+            else:
+                raise TraceDBError(
+                    f"The derived trace {self.name} presents an unknown operation {op}."
+                )
+
+            self.raw_traces[file] = self.multi * numpy.array(derived_trace)
+
+    def refine(self, time_trace={}):
+        """Fill refined_traces using evaluating the keys as member functions in Utils, numpy or scipy
+
+        The refine_traces key can be a str or a list of arguments. In case of a str, it is evaluated as a member
+        function of Utils, numpy or scipy. In case of a list, the first term is the member function while the others
+        are arguments to be fed after the traces.
+        """
+
+        for op_key in self.refined_traces.columns:
+            tr = []
+            if len(self.refined_traces[op_key]) == 1 and isinstance(
+                self.refined_traces[op_key][0], str
+            ):
+                f = open(self.refined_traces[op_key][0])
```

Comment:
You can use the `with` statement to properly close the file when not used anymore:

```python
with open(self.refined, ...) as f:
  for l in f.readlines():
    # ...
```